change Variant serde (de|)serialization

zoedberg opened the pull request:
This PR fixes `Consignment` JSON serialization (in detail its `types: TypeSystem` field).

Without this fix the serialization fails with a "key must be a string" error due to `UnionVariants` having a `NonEmptyOrdMap` with the `Variant` struct as key.

There are other possible solutions, like changing the `Union(UnionVariants<Ref>)` variant or the `UnionVariants` struct, but we would like @dr-orlovsky opinion on which solution is more appropriate.

P.S. I would like to open this on top of v2.7.2 since in RGB 0.11.1 we are still using that version, but there is no branch there (so for now I'm opening this on master)

zoedberg commented:
> 1. This PR ignores the fact that serde is not only for string serialization, but also for binary. Binary serialization must not happen through `Display`/`FromStr` and in this way the PR puts in regression to the existing functionality.

You are right, I thought we didn't care about serde binary serialization since for that we already have strict encoding, but I agree it's still better to keep that door open (note: it would have worked but it would just have been less efficient).
I fixed the PR in order to keep binary serialization efficient. This doesn't seem a backward incompatible change so please if you now agree with this PR I would ask you to create a v2.7 branch so I can change this PR base branch and then release version v2.7.3

dr-orlovsky reviewed (commented):
You just need to fix serde for `VariantName`, and leave `Variant` intact

dr-orlovsky commented:
Specifically, you need to go to `strict-encoding`, remove `serde(transparent)` on `RString` type and do manual serde implementation like here (although no need to do binary encoding there, since both the string and binary would be the same for RString)

zoedberg commented:
`Variant` is a struct and is used as a key in a map. Some serialization methods, like JSON, don't support that. How would changing the serialization of `VariantName` fix this?

dr-orlovsky commented on rust/src/util.rs:
```diff
@@ -84,13 +85,43 @@ impl Display for Sizing {
+            if serializer.is_human_readable() {
+                serializer.serialize_str(&self.to_string())
+            } else {
+                (&self.name, self.tag).serialize(serializer)
```
We need to use a proper serde struct named field serialization here not to break with previous version

zoedberg replied:
Ok, done

dr-orlovsky commented on rust/src/util.rs:
```diff
@@ -84,13 +85,43 @@ impl Display for Sizing {
+                let s = String::deserialize(deserializer)?;
+                Variant::from_str(&s).map_err(serde::de::Error::custom)
+            } else {
+                let (name, tag): (VariantName, u8) = Deserialize::deserialize(deserializer)?;
```
Ibid

zoedberg replied:
Done

dr-orlovsky commented on rust/src/util.rs:
```diff
@@ -134,7 +165,22 @@ impl Ord for Variant {
 impl Display for Variant {
     fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
-        write!(f, "{}", self.name)?;
+        write!(f, "{}:{}", self.name, self.tag)?;
```
Hm, this is actually breaking. Let me check where we use `Display` downstream (I am pretty sure it might break STON data everywhere)...

zoedberg replied:
Change reverted

dr-orlovsky commented on rust/src/util.rs:
```diff
@@ -134,7 +165,22 @@ impl Ord for Variant {
 }
+
+impl FromStr for Variant {
+    type Err = String;
```
Strings shouldn't be used as errors due to many reasons (they do not implement `Error` trait, they are not convertible into super-error types, they can't be matched etc).

zoedberg replied:
Removed the entire `FromStr` implementation since it was not symmetrical to the `Display` implementation anymore

dr-orlovsky reviewed (changes requested):
Ok, I see now, sorry. I was under impression that it is `VariantName` which is used as a key.

Anyway, pls see my comments. We also need to add roundtrip serde string and binary test in order to make sure that the change actually works,

zoedberg commented:
> Anyway, pls see my comments. We also need to add roundtrip serde string and binary test in order to make sure that the change actually works,

Comments addressed and test added.

dr-orlovsky commented on rust/src/util.rs:
```diff
@@ -84,13 +84,67 @@ impl Display for Sizing {
+        where D: Deserializer<'de> {
+            if deserializer.is_human_readable() {
+                let s = String::deserialize(deserializer)?;
+                let parts: Vec<&str> = s.split(':').collect();
```
This will really slow down parsing, since TypeSystem may have thousands and thousands of these. No need for allocation, just do 

```rust
let split = s.split(':');
let (name, tag, none) = (split.next(), split.next(), split.next()); 
if none.is_some() { return Err(...) }
```

zoedberg replied:
Ok, done

dr-orlovsky commented on rust/src/util.rs:
```diff
@@ -84,13 +84,67 @@ impl Display for Sizing {
+                    name: VariantName,
+                    tag: u8,
+                }
+                let VariantFields { name, tag } = VariantFields::deserialize(deserializer)?;
```
This will probably not work with a tagged YAML, since above in serializer you set explicit name `Variant` for serialization - and here by deriving deserializer it sets the name to `VariantFields`.

dr-orlovsky replied:
I think you can add `serde(rename = "Variant")` and this should solve the issue

zoedberg replied:
Ok, done

dr-orlovsky reviewed (changes requested):
Thank you! Two more things and seems we are done

zoedberg commented:
> Thank you! Two more things and seems we are done

Great, thank you. This should be ready, the only thing that's missing is a branch for `v2.7.2` because I need these changes on top of that and not on top of `master`. Afterwards I will open another PR to put these changes on top of `v2.8.1`

P.S. I've also bumped the MSRV since `half` now requires at least Rust 1.81.0 (you can see that from the previous CI run that was failing)

dr-orlovsky commented on Cargo.toml:
```diff
@@ -8,7 +8,7 @@ version = "2.7.2"
 homepage = "https://strict-types.org"
 repository = "https://github.com/strict-types/strict-encoding"
-rust-version = "1.77.0"
+rust-version = "1.81.0"
```
This is a breaking change which can't go into a minor release. To counter-act MSRV bump in dependency just pin it to the latest correct version in Cargo.toml (i.e. add it as a dirrect dependency with "<=" version requirement)

zoedberg replied:
Ok, reverted and fixed the dep version

dr-orlovsky commented on rust/src/util.rs:
```diff
@@ -84,13 +84,80 @@ impl Display for Sizing {
 }
 impl_strict_struct!(Variant, STRICT_TYPES_LIB; name, tag);
 
+#[cfg(feature = "serde")]
```
Let's have a comment with explanation here:

```suggestion
#[cfg(feature = "serde")]
// The manual serde implementation is needed due to `Variant` bein used as a key in maps (like enum or union fields), and serde text implementations such as JSON can't serialize map keys if they are not strings. This solves the issue, by putting string serialization of `Variant` for human-readable serializers
```

dr-orlovsky reviewed (changes requested):
Thank you!

I wish I could merge, but the reality that MSRV bump is a breaking change and the problem need to be handled differently (pls see my comment)

I have also created v2.7 branch so you can do a PR there (additionally to this one; please keep this PR against `master` and just do another one with a backport)

dr-orlovsky reviewed (approved):
ACK ac236479a49139dcbc9799fb4b2d37c9d6f797ca

dr-orlovsky commented:
Thank you.

The version was reverted, but not pinned in Cargo.toml; but (a) this is fine and (b) we can do it in some other PR if you'd like.

Merging, since it already takes forever :)

zoedberg commented:
> Thank you!
> 
> I wish I could merge, but the reality that MSRV bump is a breaking change and the problem need to be handled differently (pls see my comment)

Fixed and rebased on top of master

> I have also created v2.7 branch so you can do a PR there (additionally to this one; please keep this PR against `master` and just do another one with a backport)

You created it in strict-types but this PR is on strict-encoding, could you please create the branch here as well? Thanks

zoedberg commented:
> The version was reverted, but not pinned in Cargo.toml; but (a) this is fine and (b) we can do it in some other PR if you'd like.

Not sure what you mean, I have fixed the unwanted MSRV by adding `half = "<2.5.0"` to the dev-dependencies
